Make the logger backend more robust

larskuhtz opened the pull request:
- [x] guarantee termination by adding an (optional) timeout to `releaseLogger`.
- [x] add respawn logic and configurable exception handlers for backend. Decide whether this
     should be configurable for the user of the logger, the user of an backend, or only for the author 
     of a backend.
- [x] investigate if respawn logic and configurable handlers are needed for the logger itself.
- [x] add means to log errors in log-system to an alternate sink (e.g. stderr).

Commit:
add timeout to realseLogger

Commit:
A more robust execution model for the backend function

* The logger can raise exceptions when it can't recover from exceptions
* The backend can trigger an immediate exit (with exception) of the logger
* There are now parameters in LoggerConfig that specify how to deal with exceptions in the backend
* The LoggerException type has two no constructors for the case when the backend
   requests termination of the logger and when the logger exists because of too many
   backend exceptions.

jonsterling commented on src/System/Logger/Logger/Internal.hs:
```diff
@@ -221,15 +266,71 @@ loggerMissed ∷ Lens' (Logger a) (TVar Int)
+-- there only internal queue.
+--
+-- Exceptions of type 'BlockedIndefinitelyOnSTM' and 'NestedAtomically' are
+-- rethrowin immediately. Those exceptions indicate a bug in the code due to
```
`s/rethrowin/rethrown`

jonsterling commented on src/System/Logger/Logger/Internal.hs:
```diff
@@ -221,15 +266,71 @@ loggerMissed ∷ Lens' (Logger a) (TVar Int)
+-- Exceptions of type 'BlockedIndefinitelyOnSTM' and 'NestedAtomically' are
+-- rethrowin immediately. Those exceptions indicate a bug in the code due to
+-- unsafe usage of 'createLogger'. This exceptions shouldn't be possible when
+-- 'withLogger' is used to provide the logger and and the reference to the
```
`s/and and/and`

jonsterling commented on src/System/Logger/Types.hs:
```diff
@@ -255,20 +273,30 @@ logMsgScope = lens _logMsgScope $ \a b → a { _logMsgScope = b }
--- While this can be done in a closure of this function a more direct
--- support might be desirable.
+-- TODO there may be scenarios where chunked processing is beneficial. While
+-- this can be done in a closure of this function a more direct support might
```
`s/ a more/, more`

jonsterling commented on src/System/Logger/Types.hs:
```diff
@@ -255,20 +273,30 @@ logMsgScope = lens _logMsgScope $ \a b → a { _logMsgScope = b }
 
 -- | This is given to logger when it is created. It formats and delivers
--- individual log messages synchronously.
+-- individual log messages synchronously. The backend is called once ofr each
```
`s/ofr/for`

Commit:
Logger: add missing lens exports

Commit:
Logger: expose exit timeout in LoggerConfig

Commit:
Logger.Internal: remove redundant import

Commit:
Logger: add withLogFunction_

This function allows to provide an alternate text-based backend for logging
of errors in the logging system.

Commit:
Fix spelling of comments

Commit:
Add a test-suite

Commit:
Logger.Internal: fix 'ThreadKilled' exception in 'releaseLogger'

Commit:
Logger.Internal: fix a typo in a comment

Commit:
[test] remove debugging code

Commit:
[test] NoBackend: test with different exit timeouts

Commit:
Merge branch 'master' into more-robust-backend

Commit:
[test] NoBackend: fix module name in haddock comment

Commit:
[test] NoBackend: fix a comment

Commit:
[test] add TastyTools with a better testCaseSteps implementation

Commit:
Merge branch 'master' into more-robust-backend

Commit:
Logger.Internal: warn if messages are discarded on termination

Commit:
[test] NoBackend: give no recover tests more time to raise expected exception

jonsterling commented:
Now that some of the strange bugs have been worked out (even though the solutions to them are perhaps not super illuminating), I think that this PR is fine to merge and contains many improvements. :+1:
